instrumentation/elasticsearch/elasticsearch-rest-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/elasticsearch/rest/ElasticsearchClientAttributeExtractor.java

```diff
@@ -0,0 +1,183 @@
+
+    if (SemconvStability.emitOldHttpSemconv()) {
+      internalSet(attributes, SemanticAttributes.HTTP_URL, fullUrl);
+    }
+  }
+
+  @Override
+  public void onStart(
+      AttributesBuilder attributes, Context parentContext, ElasticsearchRestRequest request) {
+    internalSet(attributes, SemanticAttributes.HTTP_METHOD, request.getMethod());
+    setPathPartsAttributes(attributes, request);
+  }
+
+  @Override
+  public void onEnd(
+      AttributesBuilder attributes,
+      Context context,
+      ElasticsearchRestRequest request,
+      @Nullable Response response,
+      @Nullable Throwable error) {
+    if (response != null) {
+
+      setUrlAttribute(attributes, response);
+      setServerAttributes(attributes, response);
+    }
+  }
+
+  /**
+   * Identifies the corresponding URL route by matching the actual URL path against valid URL routes
+   * for that corresponding endpoint definition.
+   *
+   * <p>Once the correct URL route is found, this method retrieves the URL path parameters from the
+   * URL path by using the corresponding URL route regex with embedded named capture groups for the
+   * URL path parameters.
+   */
+  private void setPathPartsAttributes(
+      AttributesBuilder attributes, ElasticsearchRestRequest request) {
+    ElasticsearchEndpointDefinition endpointDef = request.getEndpointDefinition();
+    if (endpointDef == null) {
+      return;
+    }
+    String[] availableRoutes = endpointDef.getRoutes();
+    if (availableRoutes == null || availableRoutes.length == 0) {
+      return;
+    }
+    String urlPath = request.getEndpoint();
+    boolean routeFound = false;
+    for (int i = 0; i < availableRoutes.length && !routeFound; i++) {
+      String route = availableRoutes[i];
+      if (route.contains("{")) {
+        EndpointPattern endpointPattern = getEndpointPattern(route);
+        Matcher matcher = endpointPattern.createMatcher(urlPath);
+        if (matcher.find()) {
+          for (String key : endpointPattern.getPathPartNames()) {
+            try {
+              String value = matcher.group(key);
+              String attributeKey = "db.elasticsearch.path_parts." + key;
+              internalSet(attributes, AttributeKey.stringKey(attributeKey), value);
+            } catch (RuntimeException e) {
+              // ignore
+            }
+          }
+
+          routeFound = true;
+        }
+      }
+    }
+  }
+
+  /**
+   * For a given route creates, compiles and caches a regular expression pattern and retrieves a set
+   * of pathPartNames (names of the URL path parameters).
+   *
+   * <p>The regex pattern is later being used to match against a URL path to retrieve the URL path
+   * parameters for that route pattern using named regex capture groups.
+   *
+   * @param route The route to create the regex pattern for.
+   * @return The EndpointPattern containing the regex pattern and the URL path parameter names/keys.
+   */
+  private EndpointPattern getEndpointPattern(String route) {
```

Comment:
Instead of synchronizing you could use `computeIfAbsent`
```java
  private EndpointPattern getEndpointPattern(String route) {
    return endpointPatternMap.computeIfAbsent(route, (r) -> {
      // convert route to regex by replacing {foo} with (?<foo>[^/]+)
      String regexStr = '^' + route.replace("{", "(?<").replace("}", ">[^/]+)") + '$';
      Pattern regexPattern = Pattern.compile(regexStr);

      return new EndpointPattern(regexPattern, getPathPartNames(route));
    });
  }

  private static List<String> getPathPartNames(String route) {
    if (!route.contains("{")) {
      return Collections.emptyList();
    }
    List<String> pathPartNames = new ArrayList<>();
    Matcher matcher = pathPartNamesPattern.matcher(route);
    while (matcher.find()) {
      pathPartNames.add(matcher.group(1));
    }
    return pathPartNames;
  }
```
Another alternative would be to precompute all of this in `ElasticsearchEndpointMap` and instead of passing around route names pass around an object that can deal with this